timemachine/fe/free_energy.py

```diff
@@ -502,3 +496,122 @@ def run_sequential_sims_given_initial_states(
+    frames, boxes = sample(initial_state, md_params, max_buffer_frames=1000)
+    bound_impls = [p.bound_impl(np.float32) for p in initial_state.potentials]
+    assert initial_state.barostat is None or initial_state.barostat.temperature == temperature
+    batch_u_fns = get_batch_u_fns(bound_impls, temperature)
+    return EnergyDecomposedState(frames, boxes, batch_u_fns)
+
+
+def compute_bar_error(u_kln: NDArray) -> float:
+    assert u_kln.ndim == 3
+    assert u_kln.shape[:2] == (2, 2)
+    w_f = u_kln[1, 0] - u_kln[0, 0]
+    w_r = u_kln[0, 1] - u_kln[1, 1]
+    _, df_err = bar_with_bootstrapped_uncertainty(w_f, w_r)
+    return df_err
+
+
+def run_sims_with_greedy_bisection(
+    initial_lambdas: Sequence[float],
+    make_initial_state: Callable[[float], InitialState],
+    md_params: MDParams,
+    n_bisections: int,
+    temperature: float,
+    verbose: bool = True,
+) -> Tuple[List[Tuple[List[InitialState], NDArray]], List[StoredArrays], List[NDArray]]:
+    r"""Starting from a specified lambda schedule, successively bisect the lambda interval between the pair of states
+    with the largest BAR :math:`\Delta G` error and sample the new state with MD.
+
+    Parameters
+    ----------
+    initial_lambdas: sequence of float, length >= 2, monotonically increasing
+        Initial protocol; starting point for bisection.
+
+    make_initial_state: callable
+        Function returning an InitialState (i.e., starting point for MD) given lambda
+
+    md_params: MDParams
+        Parameters used to simulate new states
+
+    n_bisections: int
+        Number of bisection steps to perform
+
+    temperature: float
+        Temperature in K
+
+    verbose: bool
+        Whether to print diagnostic information
+
+    Returns
+    -------
+    results: list of tuples of (list of InitialState, NDArray)
+        Results from each iteration of bisction. The first element of the nth tuple is the list of InitialState objects
+        specifying the protocol after the nth iteration of bisection; the second element is a (L-1, C, 2, 2, F) array of
+        u_kln matrices for each pair of adjacent windows, where L is the number of lambda windows, C is the number of
+        energy components, and F is the number of frames per window.
+
+    frames: list of StoredArrays
+        Frames from the final iteration of bisection. Shape (L, F, N, 3) where N is the number of atoms.
+
+    boxes: list of NDArray
+        Boxes from the final iteration of bisection. Shape (L, F, 3, 3).
+    """
+
+    assert len(initial_lambdas) >= 2
+    assert np.all(np.diff(initial_lambdas) > 0), "initial lambda schedule must be monotonically increasing"
+
+    cache = lru_cache(maxsize=None)
+
+    get_initial_state = cache(make_initial_state)
+
+    @cache
+    def get_state(lamb: float) -> EnergyDecomposedState[StoredArrays]:
+        initial_state = get_initial_state(lamb)
+        return sample_state(initial_state, md_params, temperature)
```

Comment:
Good point; I didn't see any issues for the systems I tested (tyk2 and a subset of hif2a with up 48 windows), but it seems ideal to avoid caching the bound impls, since they can be reconstructed relatively quickly.

Reply:
I think this should be addressed by https://github.com/proteneer/timemachine/pull/959/commits/3c890a2c54546dfed3914b5d79dd1f9417790b59. This separates `sample_state` into `get_state(float) -> EnergyDecomposedState` and `get_samples(float) -> Tuple[StoredArrays, NDArray]` and caches only the latter.